### PR TITLE
Fix `is_euler_pseudoprime`

### PR DESCRIPTION
**Pull request by haru-44**
There are several things called Euler pseudoprime. However, the current implementation is different from the docstring description and wikipedia. So we modified the implementation to be in line with the docstring definition.

#### Release Notes

* ntheory
  * The behavior of `is_euler_pseudoprime` has changed. Composite numbers that used to return `False` may now also return `True`.

### CODE REVIEW

**Comment by smichr** [comment]
Any idea what version of definition the original code was giving?

File: sympy/ntheory/primetest.py [hunk 1]
```diff
@@ -6,12 +6,14 @@
...
 from sympy.utilities.misc import as_int
 
 
-def is_euler_pseudoprime(n, b):
-    """Returns True if n is prime or an Euler pseudoprime to base b, else False.
+def is_euler_pseudoprime(n, a):
+    """Returns True if n is prime or an Euler pseudoprime to base a, else False.
 
     Euler Pseudoprime : In arithmetic, an odd composite integer n is called an
     euler pseudoprime to base a, if a and n are coprime and satisfy the modular
```
Comment (smichr) [inline sympy/ntheory/primetest.py:19]:
```suggestion
    """Returns True if n is prime or is an odd composite integer n that is
    coprime to b and satisfies the modular
```

File: sympy/ntheory/primetest.py [hunk 2]
```diff
@@ -6,12 +6,14 @@
...
-from sympy.external.gmpy import gmpy as _gmpy, jacobi, is_square as gmpy_is_square, bit_scan1
+from sympy.external.gmpy import (gmpy as _gmpy, gcd, jacobi,
+                                 is_square as gmpy_is_square,
+                                 bit_scan1)
 from sympy.utilities.misc import as_int
 
 
-def is_euler_pseudoprime(n, b):
-    """Returns True if n is prime or an Euler pseudoprime to base b, else False.
+def is_euler_pseudoprime(n, a):
```
Comment (smichr) [inline sympy/ntheory/primetest.py:15]:
```suggestion
def is_euler_pseudoprime(n, b):
```

**Comment by smichr** [comment]
Wouldn't it be better to exclude primes from return True? Although the docstring said otherwise, it seems the tests were structured in that way, showing the first Euler pseudoprime for given bases, not the first number that would pass the test.

**Comment by haru-44** [comment]
> Any idea what version of definition the original code was giving?

Definition of Euler pseudoprime

1. $a^{\frac{n-1}{2}} \overset{?}{\equiv} \pm 1 \pmod{n}$ : [wikipedia](https://en.wikipedia.org/wiki/Euler_pseudoprime), [oeis](https://oeis.org/A006970)
2. $a^{\frac{n-1}{2}} \overset{?}{\equiv} \left( \frac{a}{n} \right) \pmod{n}$ : [gmpy2](https://gmpy2.readthedocs.io/en/latest/advmpz.html#gmpy2.is_euler_prp) (Oh, the gmpy2 documentation is wrong!)

However, the current implementation does not fall into either of these cases. The current implementation runs the Miller-Rabin primality test twice, the first time calling `mr` and the second time implementing it on its own. At least the second one is unnecessary.

> Wouldn't it be better to exclude primes from return True?

Is the question whether to include or exclude prime numbers in the definition of pseudoprime? As noted on [wikipedia,](https://en.wikipedia.org/w/index.php?title=Pseudoprime&oldid=1152955578) both definitions are possible.

> Some sources use the term pseudoprime to describe all probable primes, both composite numbers and actual primes.

When considering composite numbers that are mistakenly determined to be prime, it is rational not to include prime numbers in pseudoprime.
On the other hand, in programming for prime number determination, it is more natural to include primes in pseudoprime (e.g., `gmpy2.is_**_prp`).
So I think the name as it is now will not cause any misunderstanding, and the explanation in the docstring will make it clearer.

**Comment by smichr** [comment]
What do you think about handling the case

```python
if a == 1:
    return n % 2 and not isprime(n)  # odd composite
```

**Comment by haru-44** [comment]
When considering `is_euler_pseudoprime` as a probabilistic algorithm for determining prime numbers, the use of `isprime` internally raises the classic dilemma of the chicken and the egg. Consequently, I do not advocate the utilization of `isprime`.

If `a==1` should not be an error, then it is `True` when `n` is odd.  So, I think this can be translated into the question "Should `a==1` be treated as an error?" There is no reason to actively make it an error, so I will modify it so that it should not be an error.

File: sympy/ntheory/primetest.py [hunk 3]
```diff
@@ -6,51 +6,64 @@
...
     """
-
-    if not mr(n, [b]):
+    n, a = as_int(n), as_int(a)
+    if a < 1:
+        raise ValueError("a should be an integer greater than 0")
+    if n < 1:
+        raise ValueError("n should be an integer greater than 0")
+    if n == 1:
         return False
```
Comment (smichr) [inline sympy/ntheory/primetest.py:61]:
```suggestion
        return False
    if a == 1:
        return bool(n % 2)  # only True for odd (prime or composite)
```

Comment (smichr) [reply sympy/ntheory/primetest.py:61]: In keeping with docstring, we don't want to return True for all n when `a==1`; modification suggested.

Comment (haru-44) [reply sympy/ntheory/primetest.py:61]: Even with the current RP, when `a=1`, the result is as expected. Conversely, modifying it that way would result in `False` when `n=2, a=1`.

Comment (smichr) [reply sympy/ntheory/primetest.py:61]: I updated the suggestion so True is returned for any prime (all odd except for 2) or odd composite. The docstring indicates that we should not be returning True for even numbers that are not prime (which the currrent code does, I believe).

Comment (haru-44) [reply sympy/ntheory/primetest.py:61]: To clarify and avoid any misunderstandings, the requirement for n to be odd is to calculate $(n-1)/2$. Therefore, it is necessary to handle even numbers separately before calculating $(n-1)/2$. It is rational to consider 2 as a prime number and classify all other even numbers as not prime.

**Comment by haru-44** [comment]
thanks

**Pull request merged**